central/image/service/service_impl.go

```diff
@@ -462,8 +471,15 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 	return internalScanRespFromImage(img), nil
 }
 
-func (s *serviceImpl) acquireScanSemaphore() error {
-	if err := s.internalScanSemaphore.Acquire(concurrency.AsContext(concurrency.Timeout(maxSemaphoreWaitTime)), 1); err != nil {
+func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
+	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
+	defer cancel()
+	if err := s.internalScanSemaphore.Acquire(semaphoreCtx, 1); err != nil {
+		// If the context was canceled, we do not want to indicate the client to retry.
+		if errors.Is(err, context.Canceled) {
```

Comment:
Do we only want to retry upon `context.DeadlineExceeded` or any error that isn't `context.Canceled` is ok?

Reply:
I think in this case the only other possible error is `context.DeadlineExceeded` because the semaphore only returns `ctx.Err()` as an error. My thought was to opt-out of retries explicitly for known non-retryable errors like `context.Canceled`. If `Acquire()` ever returns some other errors, I think it's probably safer to keep retrying by default.

However, now that I think about it, we shall return gRPC status `codes.Canceled` here instead of just wrapping the error.